projects/anvil_tdr_ingest/clone_anvil_workspace.py

```diff
@@ -0,0 +1,101 @@
+"""Clone AnVIL Terra workspace and paste link of source workspace in dashboard of cloned workspace.
+
+Usage:
+    > python3 clone_anvil_workspace.py -sw [src_workspace] -sn [src_namespace] -dw [dest_workspace] -dn [dest_namespace] [-ads]"""
+
+import argparse
+import json
+import pandas as pd
+import requests
+
+from utils import check_workspace_exists, clone_workspace, \
+    update_workspace_dashboard, make_create_workspace_request, \
+    get_workspace_authorization_domain
+
+def create_update_dashboard_message(attribute_name, attribute_value):
+    """Return request string for a single operation to create a non-array attribute."""
+
+    return '[{"op":"AddUpdateAttribute","attributeName":"' + attribute_name + '", "addUpdateAttribute":"Source Workspace URL: ' + attribute_value + '"}]'
+
+
+def check_clone_workspace_exists(dest_namespace, dest_workspace):
+    """Check if a Terra workspace with user input name for clone already eixsts."""
+
+    # check workspace exists
+    ws_exists, ws_exists_message = check_workspace_exists(dest_workspace, dest_namespace)
+
+    # if workspace exists
+    if ws_exists:
+        print(f"Workspace already exists with name: {dest_namepsace}/{dest_workspace}.")
+        print(f"Please choose unique name or delete existing workspace and try again.")
+        print(f"Existing workspace details: {json.dumps(json.loads(ws_exists_message), indent=2)}")
+        return True, ws_exists_message
+
+    ## workspace doesn't exist (404), create workspace
+    return False, ws_exists_message
+
+
+def setup_anvil_workspace_clone(src_namespace, src_workspace, dest_namespace, dest_workspace, auth_domains=None):
+    """Clone workspace and update dashboard with url to source workspace."""
+
+    print(f"Starting clone of workspace.")
+
+    # check if workspace with name already exists
+    clone_exists, clone_exists_message = check_clone_workspace_exists(dest_namespace, dest_workspace)
+
+    # workspace exists --> prompt user to pick mew name and exit script
+    if clone_exists:
+        return
+
+    # workspace does not exist --> start set up for cloning workspace
+    # get auth domains of src workspace
+    src_auth_domains = get_workspace_authorization_domain(src_workspace, src_namespace)
+
+    # src workspace ADs, user input ADs
```

Comment:
minor, but could be nice to put this into a separate function `format_auth_domains(src_auth_domains, input_auth_domains)` - that way you could in theory unit test it, and this would be easier to read